keps/140-scenario-based-simulation/README.md

```diff
@@ -0,0 +1,397 @@
+The scenario will continue to run until all events in .spec.Events are completed, and when all events are completed, the scenario will be "Paused" phase in the case .spec.Events doesn't have "Done" operation.
+
+So, it is strongly recommended adding events to running Scenario only after Scenario have reached "Paused" phase. 
+(since FakeTime has stopped moving forward in "Paused" phase as described above.)
+Otherwise you may add the past FakeTime events and they are ignored by running Scenario.
+
+##### configure when to update scenarioResults
+
+As described in the above, the controller only update status.scenarioResults in Scenario resource when proceeding to the next FakeTime.
+
+This is because kube-apiserver will be so busy if the controller update status.scenarioResults everytime it updated,
+especially when the size of Scenario is so big.
+
+> etcd is designed to handle small key value pairs typical for metadata. Larger requests will work, but may increase the latency of other requests. By default, the maximum size of any request is 1.5 MiB. This limit is configurable through --max-request-bytes flag for etcd server.
+https://etcd.io/docs/v3.4/dev-guide/limit/#request-size-limit
+
+And, sometimes, users may want to reduce this updating request more for some reasons. 
+
+For example, 
+- when watching Scenario for dynamic event addition 
+- when using Scenario for accurate benchmark testing, users may want to reduce the request to update Scenario for kube-apiserver as much as possible
+
+We can add a new configuration environment variable `UPDATE_SCENARIO_RESULTS_STRATEGY` and define some strategy like:
+- `UPDATE_SCENARIO_RESULTS_STRATEGY=AtMovingNextTime`: default value. update status.scenarioResults in Scenario resource when proceeding to the next FakeTime.
+- `UPDATE_SCENARIO_RESULTS_STRATEGY=OnPause`: update status.scenarioResults in Scenario resource when the Scenario's phase is `Paused` or `Failed`.
+- `UPDATE_SCENARIO_RESULTS_STRATEGY=OnDone`: update status.scenarioResults in Scenario resource when the Scenario's phase is `Done` or `Failed`.
+
+#### The result calculation packages
+
+ScenarioResults only has the simple data that represent what was happen during the scenario.
+
+So, we will provide useful functions and data structures to analize the result. 
+
+For example:
+- the function to aggregate changes in allocation rate of the entire cluster.
+- the function to aggregate changes in resource utilization for each Node.
+- the function to aggregate data by Pod.
+- (Do you have any other idea? Tell us!)
+
+By putting only the minimum simple information in ScenarioResults and providing functions to change it into a user-friendly structs, many data structures can be supported in the future without any changes to API.
+
+#### Prohibitions and Restrictions
+
+When Scenario is created, the scenario is started by the controller. The scenario is run one by one, and multiple scenarios are never run at the same time.
```

Comment:
Does "one by one" mean in the order of `metadata.CreationTimestamp`?? or other?

Reply:
> Does "one by one" mean in the order of metadata.CreationTimestamp?? 

I guess the controller catches the `Scenario` creation with EventHandler or Watch, so maybe Yes?

Reply:
> so maybe Yes?

I think it depends on the implementation of the Scenario reconciler.

Assume one Scenario is running and several pending Scenarios exist, when the running Scenario got completed, which Scenario did the reconciler pick up for the next startup??